### DISCUSSION TITLE
Infinite Carousel

### DISCUSSION THREAD

**Discussion by wagnerbit** [CommunityToolkit/Labs-Windows#498] [Ideas]
Hey guys! I hope you are all well.
I would like your help to solve a problem I am having with a control (Carousel). Is there any way to make the carousel behave like infinity? An example is the Microsoft Store carousel, where the user can navigate infinitely left or right.
I will be very grateful if someone can help me.

![image](https://github.com/CommunityToolkit/WindowsCommunityToolkit/assets/11899700/237e43cb-b815-4b02-9c2f-ebc7fde9c7cb)

**Comment by michael-hawker** [comment]
Hi @wagnerbit this isn't something that's built into the existing Toolkit carousel control as far as I know. @JustinXinLiu would know more about the store Carousel and how the infinite scrolling works; we've worked with his team to bring other Store controls to the Toolkit - we've been eyeing the Carousel as well for a future collaboration. 😊 That's partly why we didn't bring the existing Carousel forward in 8.0 as it's one of the older implementations in the Toolkit, so it makes sense to get an updated version together instead.

Moving this to labs as a placeholder to track this request, as I know there's been other requests for Carousel already from the release:
- https://github.com/CommunityToolkit/Windows/discussions/228#discussioncomment-6965258
- https://github.com/CommunityToolkit/Windows/discussions/228#discussioncomment-6965358
- https://github.com/CommunityToolkit/Windows/discussions/232#discussion-5614538

**Reply by wagnerbit** [reply]
Thank you very much for your feedback @michael-hawker

**Reply by michael-hawker** [reply]
Someone was asking about this again, wanted to call out too that @JustinXinLiu had a small pre-cursor implementation in a sample app to that if anyone wanted to take it as a starting point. That code is here: https://github.com/microsoft/InsiderDevTourDemos19/blob/master/Sessions/ui/Uwp.App/Uwp.App/Layouts/VirtualizingUniformCarousalStackLayout.cs (The original Carousel in the toolkit was written long ago and a bit cobbled together, so it's something we were hoping to just update forwards with something else.)

**Comment by ghost1372** [comment]
@michael-hawker  any news?

**Comment by yourordinarycat** [comment]
Funny enough, there's been a basic "infinite carousel" control in Windows XAML for ages now - `LoopingSelector`. Even more funny, this control is sealed, has no constructor, is only there for a few controls that need it (`TimePickerFlyout` in particular), and the API would be kinda busted for a standard Panel (though that's par for the course for something that only gets used internally): https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.loopingselector?view=winrt-22621

Lucky for us, this control is included in the WinUI 3 source code - look for `LoopingSelector_Partial`. The code could be ported to C# and adapted to better support its use as a XAML `Panel`, thoughts on this?

**Reply by ghost1372** [reply]
there is no xaml codes for loopingselector

**Reply by yourordinarycat** [reply]
@ghost1372 Yes, I'm aware of that specific control not being usable currently - it's only there for internal use by a few controls (such as `TimePickerFlyout`), and its API doesn't make it suitable for a XAML Panel.

What I was saying is, the source code for that specific control is out there thanks to WinUI 3's open sourcing, so the toolkit could build its own solution based on that code, making any necessary adaptations for the API surface and code. Apologies if that part wasn't clear before.

**Reply by ghost1372** [reply]
Hi @yourordinarycat 
Today I found out that Uno has ported the LoopingSelector, and it works fine in their TimePicker and other controls. However, when I copy-pasted their code into a blank WinUI app (replacing some namespaces and fixing errors), it runs with some issues.
I couldn’t fix the problem. Can you take a look?
As you can see, the selection doesn’t update after changing items, but moving the mouse out and back in updates the selection. If the selection isn’t updated correctly, the centered item doesn’t behave properly either.

![ss](https://github.com/user-attachments/assets/da091f04-c7be-480f-aa05-63845f481858)

Here you can find the Uno C# source code:


https://github.com/unoplatform/uno/tree/master/src/Uno.UI/UI/Xaml/Controls/Primitives/LoopingSelector

**Reply by ghost1372** [reply]
download my sample
[App2.zip](https://github.com/user-attachments/files/23281947/App2.zip)

**Comment by Love-Hitek** [comment]
Adding another voice to having this added.

**Comment by keithnicholson** [comment]
At a minimum, can it at least be added as a victim to the document on what has been lost?
